### PR TITLE
added OpenShift template for Quay app

### PR DESCRIPTION
**Pull request by tparikh**
### Description of Changes

Added OpenShift template to setup Quay.io application on OSD cluster. This template will be used by App-SRE CI/CD pipeline to install & update Quay application on OSD cluster.

#### Issue

https://issues.jboss.org/browse/PROJQUAY-11

## Reviewer Checklist

- [ ] It works!
- [ ] Comments provide sufficient explanations for the next contributor
- [ ] Tests cover changes and corner cases
- [ ] Follows Quay syntax patterns and format

Signed-off-by: Tejas Parikh <tparikh@redhat.com>

### CODE REVIEW

**Comment by tparikh** [comment]
OpenShift objects created from the template by running `oc process -f quay-app.yaml -o yaml` command:

```yaml
apiVersion: v1
items:
- apiVersion: rbac.authorization.k8s.io/v1beta1
  kind: Role
  metadata:
    name: quay
  rules:
  - apiGroups:
    - ""
    resources:
    - secrets
    verbs:
    - get
    - patch
    - update
  - apiGroups:
    - ""
    resources:
    - namespaces
    verbs:
    - get
  - apiGroups:
    - extensions
    - apps
    resources:
    - deployments
    verbs:
    - get
    - list
    - patch
    - update
    - watch
- apiVersion: rbac.authorization.k8s.io/v1beta1
  kind: RoleBinding
  metadata:
    name: quay
  roleRef:
    apiGroup: rbac.authorization.k8s.io
    kind: Role
    name: quay
  subjects:
  - kind: ServiceAccount
    name: default
- apiVersion: v1
  kind: Service
  metadata:
    name: quay-clusterip-service
  spec:
    ports:
    - name: clusterip
      port: 443
      protocol: TCP
      targetPort: 8443
    selector:
      quay-component: app
    type: ClusterIP
- apiVersion: v1
  kind: Service
  metadata:
    name: quay-loadbalancer-service
  spec:
    loadBalancerIP: null
    ports:
    - name: loadbalancer
      port: 443
      protocol: TCP
      targetPort: 8443
    selector:
      quay-component: app
    type: LoadBalancer
- apiVersion: apps/v1
  kind: Deployment
  metadata:
    labels:
      quay-component: app
    name: quay-app
  spec:
    minReadySeconds: 0
    progressDeadlineSeconds: 600s
    replicas: 1
    revisionHistoryLimit: 10
    selector:
      matchLabels:
        quay-component: app
    strategy:
      rollingUpdate:
        maxSurge: 25%
        maxUnavailable: 25%
      type: RollingUpdate
    template:
      metadata:
        labels:
          quay-component: app
      spec:
        containers:
        - image: quay.io/app-sre/quay:latest
          livenessProbe:
            httpGet:
              path: /health/instance
              port: 8443
            initialDelaySeconds: 15
            periodSeconds: 30
            timeoutSeconds: 10
          name: quay-app
          ports:
          - containerPort: 8443
          readinessProbe:
            httpGet:
              path: /health/endtoend
              port: 8443
            initialDelaySeconds: 15
            periodSeconds: 30
            timeoutSeconds: 10
          resources:
            limits:
              cpu: 1
              memory: 1024Mi
            requests:
              cpu: 1
              memory: 1024Mi
          volumeMounts:
          - mountPath: /conf/stack
            name: configvolume
            readOnly: false
        imagePullSecrets:
        - name: redhat-quay-pull-secret
        volumes:
        - name: configvolume
          secret:
            secretName: quay-config-secret
kind: List
metadata: {}
```

File: deploy/openshift/quay-app.yaml [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+    displayName: loadbalancer service target port
+  - name: QUAY_APP_CONFIG_SECRET
+    value: "quay-config-secret"
+    displayName: quay app config secret
+  - name: QUAY_APP_DEPLOYMENT_REPLICAS
+    value: "1"
+    displayName: quay app deployment replicas
+  - name: QUAY_APP_MEMORY_REQUEST
+    value: "1024Mi"
+    displayName: "quay app memory request"
```
Comment (jakedt) [inline deploy/openshift/quay-app.yaml]: I think the official docs ask for 4GB for memory.

File: deploy/openshift/quay-app.yaml [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  - name: QUAY_APP_DEPLOYMENT_REVISION_HISTORY_LIMITS
+    value: "10"
+    displayName: quay app deployment revision history limits
+  - name: QUAY_APP_DEPLOYMENT_STRATEGY_TYPE
+    value: "RollingUpdate"
+    displayName: quay app deployment strategy
+  - name: QUAY_APP_DEPLOYMENT_MAX_SURGE
+    value: "25%"
+    displayName: quay app deployment max surge
+  - name: QUAY_APP_DEPLOYMENT_MAX_UNAVAILABLE
```
Comment (jakedt) [inline deploy/openshift/quay-app.yaml]: Is there a reason we don't want this set to `0`? I think 25% is too big for a prod deploy.

**Comment by jzelinskie** [comment]
Should this include some kind of documentation file? This also seems like something that could be referenced in the Red Hat product documentation.

File: deploy/openshift/quay-app.yaml (hunk 2 above)
Comment (tparikh) [reply deploy/openshift/quay-app.yaml]: @jakedt do you want to set maxUnavailable to 0 right? Should we set maxSurge to `5%`?

**Comment by tparikh** [comment]
> Should this include some kind of documentation file? This also seems like something that could be referenced in the Red Hat product documentation.

We can update the docs and point to this file if needed. One big difference between this template and existing k8s files is that the current enterprise deployment files have `namespace` hard coded.

File: deploy/openshift/quay-app.yaml (hunk 2 above)
Comment (tparikh) [reply deploy/openshift/quay-app.yaml]: I have updated maxSurge to 1 and maxUnavailable to 0. We can override these values in SAAS repo for actual deployment.
